### PR TITLE
field name compatibility for SCK

### PR DESCRIPTION
**Pull request by rockb1017**
backward compatible field names for container metadata.

### CODE REVIEW

File: helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl [hunk 1]
```diff
@@ -277,11 +277,9 @@ processors:
           tag_name: splunk.com/exclude
           from: pod
         - key: splunk.com/index
-          tag_name: com.splunk.index
-          from: namespace
-        - key: splunk.com/index
-          tag_name: com.splunk.index
           from: pod
+        - key: splunk.com/index
+          from: namespace
```
Comment (dmitryax) [inline helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl:282]: Why we need `k8s.namespace.labels.splunk.com/index` instead of setting `com.splunk.index` directly?

File: helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl [hunk 2]
```diff
@@ -302,6 +300,76 @@ processors:
      - k8s.namespace.name
      - k8s.pod.name
      - k8s.pod.uid
+  {{- else }}
+  resource/splunk:
+    attributes:
+    - key: host.name
+      from_attribute: k8s.node.name
+      action: upsert
```
Comment (dmitryax) [inline helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl:308]: `host.name` is already set by resource detection processor. Why it has to be overridden?

File: helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl [hunk 3]
```diff
@@ -302,6 +300,76 @@ processors:
...
+    - key: com.splunk.sourcetype
+      from_attribute: k8s.pod.annotations.splunk.com/sourcetype
+      action: upsert
```
Comment (dmitryax) [inline helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl:311]: This is already defined in `resource/logs` processor

File: helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl [hunk 4]
```diff
@@ -302,6 +300,76 @@ processors:
...
+    - key: com.splunk.index
+      from_attribute: k8s.namespace.annotations.splunk.com/index
+      action: upsert
+    - key: service.name
+      from_attribute: k8s.pod.name
+      action: upsert
```
Comment (dmitryax) [inline helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl:317]: Why this is needed?

File: helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl [hunk 5]
```diff
@@ -302,6 +300,76 @@ processors:
...
+  resource/splunk2:
+    attributes:
+    - key: service.name
+      from_attribute: k8s.pod.labels.app
+      action: upsert
```
Comment (dmitryax) [inline helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl:322]: this logic already exist in `resource/logs` processor

File: helm-charts/splunk-otel-collector/values.yaml [hunk 6]
```diff
@@ -329,6 +329,12 @@ logsCollection:
     excludePaths: []
     # Boolean for ingesting the agent's own log
     excludeAgentLogs: true
+    # Field name conventions to use. (Only for those who are migrating from Splunk Connect for Kubernetes helm chart)
+    fieldNameConvention:
+      # Boolean for renaming pod metadata fields to match to Splunk Connect for Kubernetes helm chart.
+      renameFieldsSck: false
+      # Boolean for keeping Otel convention fields after renaming it
+      keepOtelContention: false
```
Comment (dmitryax) [inline helm-charts/splunk-otel-collector/values.yaml:337]: This is applicable to metrics as well, right? I think it should be under `splunkPlatform` config group?

And do we really need an option to send both attributes? Maybe `nameConvention: <otel|sck>`?

File: rendered/manifests/agent-only/clusterRole.yaml [hunk 7]
```diff
@@ -1,83 +0,0 @@
----
```
Comment (dmitryax) [inline rendered/manifests/agent-only/clusterRole.yaml:1]: Why did you remove all the rendered files?

File: helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl (hunk 1 above)
Comment (rockb1017) [reply helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl:282]: in order to guarantee that pod annotation gets the precedence over namespace annotation.

File: helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl (hunk 2 above)
Comment (rockb1017) [reply helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl:308]: ah i see. great. i will remove

File: helm-charts/splunk-otel-collector/values.yaml (hunk 6 above)
Comment (rockb1017) [reply helm-charts/splunk-otel-collector/values.yaml:337]: okay I will move to splunkPlatform. 
as for the use case for both, maybe for during migration phase, we don't want to break their already set up Splunk queries and stuff, but also they want to correlate logs with data in O11y.

File: helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl (hunk 4 above)
Comment (rockb1017) [reply helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl]: I thought this is needed for O11y

File: rendered/manifests/agent-only/clusterRole.yaml (hunk 7 above)
Comment (rockb1017) [reply rendered/manifests/agent-only/clusterRole.yaml]: added them back. `make render` only deleted them, but not create new ones.

File: helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl (hunk 1 above)
Comment (dmitryax) [reply helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl:282]: It should be already guaranteed. Once attribute from pod is set [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/k8sattributesprocessor/processor.go#L123), the same attribute from namespace will be _inserted_ [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/k8sattributesprocessor/processor.go#L133) which is noop if the same key was set from pod

File: helm-charts/splunk-otel-collector/values.yaml (hunk 6 above)
Comment (dmitryax) [reply helm-charts/splunk-otel-collector/values.yaml:337]: Ok, sounds good

File: helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl (hunk 4 above)
Comment (dmitryax) [reply helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl]: Everything needed for o11y is already here

File: helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl (hunk 1 above)
Comment (dmitryax) [reply helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl:282]: We just need to flip the order here. Please submit another PR for that fix to keep git history clean

Comment (rockb1017) [reply helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl:282]: okay great. will revert it.

File: CHANGELOG.md [hunk 8]
```diff
@@ -10,6 +10,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
 - Add initContainer for log checkpoint migration from Fluentd to Otel agent (#253)
 - Add index routing for Splunk Enterprise/Cloud customers (#256)
+- Field name compatibility for SCK and index routing precedence (#258)
```
Comment (dmitryax) [inline CHANGELOG.md:13]: Please move this to _Unreleased_ section

File: helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl [hunk 9]
```diff
@@ -302,6 +302,64 @@ processors:
...
+  {{- if (eq (include "splunk-otel-collector.splunkPlatformEnabled" .) "true") }}
+  resource/splunk:
+    attributes:
+    - key: com.splunk.index
+      from_attribute: k8s.namespace.annotations.splunk.com/index
+      action: upsert
+    - key: com.splunk.index
+      from_attribute: k8s.pod.annotations.splunk.com/index
+      action: upsert
+  {{- end }}
```
Comment (dmitryax) [inline helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl:315]: This section is not needed, right?

File: helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl [hunk 10]
```diff
@@ -302,6 +302,64 @@ processors:
      - k8s.namespace.name
      - k8s.pod.name
      - k8s.pod.uid
+  {{- else }}
```
Comment (dmitryax) [inline helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl:305]: Do we need this `else` here? fluentd logs have the same otel attribute names in this chart, so I think it make sense to apply the SCK conventions uniformly

File: helm-charts/splunk-otel-collector/values.yaml (hunk 6 above)
Comment (dmitryax) [reply helm-charts/splunk-otel-collector/values.yaml:337]: Do you still plan to move it under `splunkPlatform` and apply the same renaming to metrics? Or it's needed for logs only?

Comment (dmitryax) [reply helm-charts/splunk-otel-collector/values.yaml:337]: There is also a typo: conVention

Comment (dmitryax) [reply helm-charts/splunk-otel-collector/values.yaml:337]: And please make it `true` by default, otherwise all attributes removed

File: helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl (hunk 9 above)
Comment (rockb1017) [reply helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl:315]: oh yea, you are right...

File: helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl [hunk 11]
```diff
@@ -303,6 +303,53 @@ processors:
...
+    - key: k8s.pod.name
+      action: delete
+    - key: k8s.pod.uid
+      action: delete
+    - key: k8s.namespace.name
+      action: delete
+    {{- range $_, $label := .Values.extraAttributes.podLabels }}
+    - key: {{ printf "k8s.pod.labels.%s" $label }}
+      action: delete
+    {{- end }}
```
Comment (dmitryax) [inline helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl:351]: Do we need two more processors for this? Since it's applied to logs only, we could add these attribute manipulation rules in existing `resource/logs` helper. It should give better performance and no need to copy this logic across agent and gateway.

Also then you'd need to put `resource/logs` after `resource` processor because `resource` processor adds `k8s.cluster.name` that you want to rename.

Comment (rockb1017) [reply helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl:351]: can even the `removedups` be in a single processor? it wouldn't mess up things by like deleting the attribute before inserting it?

Comment (dmitryax) [reply helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl:351]: All the attribute actions are evaluated sequentially in resource processor. So just make sure they are put in the right order

File: helm-charts/splunk-otel-collector/templates/config/_common.tpl [hunk 12]
```diff
@@ -101,6 +101,49 @@ resource/logs:
...
     {{- end }}
+    {{- if .Values.splunkPlatform.fieldNameConvention.renameFieldsSck }}
+    - key: container_name
+      from_attribute: k8s.container.name
+      action: upsert
+    - key: cluster_name
+      from_attribute: k8s.cluster.name
+      action: upsert
+    - key: container_id
+      from_attribute: k8s.container.id
```
Comment (dmitryax) [inline helm-charts/splunk-otel-collector/templates/config/_common.tpl:112]: This attribute is actually recorded as `container.id`. Please also change it below with `delete` action

File: helm-charts/splunk-otel-collector/templates/config/_common.tpl [hunk 13]
```diff
@@ -101,6 +101,49 @@ resource/logs:
...
+    - key: namespace
+      from_attribute: k8s.namespace.name
+      action: upsert
+    {{- range $_, $label := .Values.extraAttributes.podLabels }}
+    - key: {{ printf "label_%s" $label }}
+      from_attribute: {{ printf "k8s.pod.labels.%s" $label }}
+      action: upsert
+    {{- end }}
+    {{- end }}
+    {{- if not .Values.splunkPlatform.fieldNameConvention.keepOtelContention }}
```
Comment (dmitryax) [inline helm-charts/splunk-otel-collector/templates/config/_common.tpl:129]:
```suggestion
    {{- if not .Values.splunkPlatform.fieldNameConvention.keepOtelConvention }}
```

Comment (dmitryax) [reply helm-charts/splunk-otel-collector/templates/config/_common.tpl:129]: Please look at the rendered files to verify desired outcome

File: CHANGELOG.md [hunk 14]
```diff
@@ -4,6 +4,10 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
 ## Unreleased
 
+### Added
+
+- Field name compatibility for SCK and index routing precedence (#258)
```
Comment (dmitryax) [inline CHANGELOG.md:9]:
```suggestion
- Field name compatibility for SCK (#258)
```

**Review by dmitryax — COMMENTED** [review]
One last correction. Otherwise looks good

**Comment by rockb1017** [comment]
Thank you so much for detailed review!

**Review by dmitryax — APPROVED** [review]
